# Week 7: Single Nucleotide and Structural Variant Analysis

Genetic variants refer to differences that make each personâ€™s genome unique. We can identify an individual's variants by comparing the DNA sequences of the individual to the DNA sequence of a reference genome. In the case of the human and mouse genome, these reference sequences are maintained by the Genome Reference Consortium and continuously updated. In a span of 3 lectures, we will get acquainted with methods to identify these variants from DNA sequences generated using massively-parallel sequencing platforms.

## Assignment

Complete the step-by-step workflow in [Assignment.md](Assignment.md), which uses several concepts presented in the first lecture focused on single-nucleotide variants. There are no deliverables for this assignment.

## Lecture 1: Single-nucleotide variants

This lecture will focus on the detection of single nucleotide variants (SNVs). As the name suggests, our goal is to detect mutations that affect a single base in the genome. But the methods to detect small indels overlap with the methods to detect SNVs, and so we will include small indels in these analyses.

* [Slides](https://docs.google.com/presentation/d/1YHKhvvG7QVc1sggoWli9neKdadxfya15MolNh65hNqg/edit?usp=sharing)
* [Tutorial](Tutorial1.md)

## Lecture 2: Structural variant detection

This lecture will focus on detection of structural variants. We will discuss the basics of SV detection using paired-end reads, and learn the advantages and disadvantages of a few of the methods developed for it.

* [Slides](https://docs.google.com/presentation/d/18dKVBiAMO1SrmTyljD5uVJY2_0Gk7puYowC4UKFPUh0/edit?usp=sharing)
* [Tutorial](Tutorial2.md)

## Lecture 3: Miscellaneous topics 

This lecture discusses topics related to the first two lectures, including target capture experiments, variant detection in cancer genomics, the GATK workflows for germline and somatic calling, alignment-free genotyping and long-read sequencing.

